File: 프로그래머스/1/250125. ［PCCE 기출문제］ 9번 ／ 이웃한 칸/［PCCE 기출문제］ 9번 ／ 이웃한 칸.cpp
#include <string>
#include <vector>

using namespace std;

int dy[4] = {0, 0, -1, 1};
int dx[4] = {-1, 1, 0, 0};

int solution(vector<vector<string>> board, int h, int w) {
    int answer = 0;
    
    string cur_color = board[h][w];
    
    for(int i = 0 ; i < 4 ; i++) {
        int next_h = h + dy[i];
        int next_w = w + dx[i];
        if(next_h >= 0 && next_h < board.size() && next_w >= 0 && next_w < board[0].size()){
            if(cur_color == board[next_h][next_w])
                answer++;
        }
        
        
    }
    
    return answer;
}

File: 프로그래머스/1/250125. ［PCCE 기출문제］ 9번 ／ 이웃한 칸/［PCCE 기출문제］ 9번 ／ 이웃한 칸.java
import java.util.*;
import java.util.stream.*;

class Solution {
    
    int[] dy = new int[]{1,-1, 0, 0};
    int[] dx = new int[]{0, 0, 1, -1};
    
    int height;
    int weight;
    String color;
    String[][] board;
    
    public int getSameCnt(int h, int w) {
        int cnt = 0;
        for(int i = 0 ; i < 4 ; i++) {
            int ny = h + dy[i];
            int nx = w + dx[i];
            if(ny >= 0 && ny < height && nx >= 0 && nx < weight)
                if(board[ny][nx].equals(color))
                    cnt++;
        }
        return cnt;
    }
    
    public int solution(String[][] board, int h, int w) {
        this.board = board;
        this.height = board.length;
        this.weight = board[0].length;
        this.color = board[h][w];
        
        return getSameCnt(h, w);
    }
}